substrate/primitives/runtime-interface/src/pass_by.rs

```diff
@@ -15,419 +15,667 @@
-///
-/// # Example
-/// ```
-/// # use sp_runtime_interface::pass_by::{PassBy, Enum};
-/// #[derive(Clone, Copy)]
-/// enum Test {
-///     Test1,
-///     Test2,
-/// }
-///
-/// impl From<Test> for u8 {
-///     fn from(val: Test) -> u8 {
-///         match val {
-///             Test::Test1 => 0,
-///             Test::Test2 => 1,
-///         }
-///     }
-/// }
+/// Raw FFI type: `u32` (a pointer to the byte blob)
+pub struct AllocateAndReturnPointer<T, const N: usize>(PhantomData<(T, [u8; N])>);
+
+impl<T, const N: usize> RIType for AllocateAndReturnPointer<T, N> {
+	type FFIType = u32;
+	type Inner = T;
+}
+
+#[cfg(not(substrate_runtime))]
+impl<T, const N: usize> IntoFFIValue for AllocateAndReturnPointer<T, N>
+where
+	T: AsRef<[u8]>,
+{
+	fn into_ffi_value(value: Self::Inner, context: &mut dyn FunctionContext) -> Result<u32> {
+		let value = value.as_ref();
+		let addr = context.allocate_memory(value.len() as u32)?;
+		context.write_memory(addr, value)?;
+		Ok(addr.into())
+	}
+}
+
+#[cfg(substrate_runtime)]
+impl<T: codec::Decode, const N: usize> FromFFIValue for AllocateAndReturnPointer<T, N>
+where
+	T: From<[u8; N]>,
+{
+	fn from_ffi_value(arg: Self::FFIType) -> Self::Inner {
+		let value = unsafe { Vec::from_raw_parts(arg as *mut u8, N, N) };
+		let array = unsafe { *(value.as_ptr() as *const [u8; N]) };
+		T::from(array)
+	}
+}
+
+/// (DEPRECATED) Return `T` as a blob of bytes into the runtime.
 ///
-/// impl TryFrom<u8> for Test {
-///     type Error = ();
+/// Uses `T::AsRef<[u8]>` to cast `T` into a `&[u8]`, allocates runtime memory
+/// using the legacy allocator, copies the slice into the runtime memory, and
+/// returns a pointer to it.
 ///
-///     fn try_from(val: u8) -> Result<Test, ()> {
-///         match val {
-///             0 => Ok(Test::Test1),
-///             1 => Ok(Test::Test2),
-///             _ => Err(()),
-///         }
-///     }
-/// }
+/// THIS STRATEGY IS DEPRECATED; DO NOT USE FOR NEW HOST FUNCTIONS!
 ///
-/// impl PassBy for Test {
-///     type PassBy = Enum<Self>;
-/// }
-/// ```
-pub struct Enum<T: Copy + Into<u8> + TryFrom<u8>>(PhantomData<T>);
+/// Raw FFI type: `u64` (a fat pointer; upper 32 bits is the size, lower 32 bits is the pointer)
+pub struct AllocateAndReturnFatPointer<T>(PhantomData<T>);
 
-#[cfg(feature = "std")]
-impl<T: Copy + Into<u8> + TryFrom<u8>> PassByImpl<T> for Enum<T> {
-	fn into_ffi_value(instance: T, _: &mut dyn FunctionContext) -> Result<Self::FFIType> {
-		Ok(instance.into() as u32)
-	}
+impl<T> RIType for AllocateAndReturnFatPointer<T> {
+	type FFIType = u64;
+	type Inner = T;
+}
 
-	fn from_ffi_value(_: &mut dyn FunctionContext, arg: Self::FFIType) -> Result<T> {
-		T::try_from(arg as u8).map_err(|_| format!("Invalid enum discriminant: {}", arg))
+#[cfg(not(substrate_runtime))]
+impl<T> IntoFFIValue for AllocateAndReturnFatPointer<T>
+where
+	T: AsRef<[u8]>,
+{
+	fn into_ffi_value(value: Self::Inner, context: &mut dyn FunctionContext) -> Result<u64> {
```

Comment:
Nit: Just to be consistent with the other impls.
```suggestion
	fn into_ffi_value(value: Self::Inner, context: &mut dyn FunctionContext) -> Result<Self::FFIType> {
```